### DISCUSSION TITLE
Why _castToPure in console.sol can convert a view method into pure method?

### DISCUSSION THREAD

**Discussion by passer-byzhang** [NomicFoundation/hardhat#5116] [Troubleshooting] [answered]
Hi folks, I have a question when I read the console.sol. There is a method which can convert a view method into a pure one. Will there be any changes to this method in the process？ Is there any solidity feature involved in this? and where can I learn about it? 
```
    function _castToPure(
      function(bytes memory) internal view fnIn
    ) internal pure returns (function(bytes memory) pure fnOut) {
        assembly {
            fnOut := fnIn
        }
    }
```

**Comment by kanej** [comment, marked as answer]
> Will there be any changes to this method in the process？

Perhaps it uses a bit more gas (I don't know if the compiler will optimize this), but otherwise there is no functionality change.

> Is there any solidity feature involved in this?

No, it's assembly hackery.

> and where can I learn about it?

https://twitter.com/devtooligan/status/1654945451584667648
https://gist.github.com/devtooligan/9031f859469138307d0976455b7ddffa
